pkg/distributor/otlp/cmd/generate/main.go

```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package main
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/alecthomas/kingpin/v2"
+
+	// Dummy dependency to ensure these files get vendored and we can use them as transformation sources.
+	_ "github.com/prometheus/prometheus/storage/remote/otlptranslator/prometheusremotewrite"
+)
+
+func main() {
+	app := kingpin.New("generate", "Generate OTLP translation code.")
+
+	generate := func(*kingpin.ParseContext) error {
+		sources, err := copySources()
+		if err != nil {
+			return err
+		}
+		if err := patchSources(sources); err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	app = app.Action(generate)
+	kingpin.MustParse(app.Parse(os.Args[1:]))
+}
+
+func patchSources(sources []string) error {
+	if out, err := exec.Command("gopatch", "-p", "cmd/generate/mimirpb.patch", "./").CombinedOutput(); err != nil {
+		return fmt.Errorf("failed to execute gopatch: %s", out)
+	}
+	sed := "sed"
+	if runtime.GOOS == "darwin" {
+		sed = "gsed"
+	}
+	for _, fname := range sources {
+		if out, err := exec.Command(sed, "-i", "s/PrometheusConverter/MimirConverter/g", fname).CombinedOutput(); err != nil {
+			return fmt.Errorf("failed to execute sed: %s", out)
+		}
+		if out, err := exec.Command(sed, "-i", "s/Prometheus remote write format/Mimir remote write format/g", fname).CombinedOutput(); err != nil {
+			return fmt.Errorf("failed to execute sed: %s", out)
+		}
+		if out, err := exec.Command("goimports", "-w", "-local", "github.com/grafana/mimir", fname).CombinedOutput(); err != nil {
+			return fmt.Errorf("failed to execute goimports: %s", out)
+		}
+	}
+
+	return nil
+}
+
+func copySources() ([]string, error) {
+	dpath := filepath.Join("..", "..", "..", "vendor", "github.com", "prometheus", "prometheus", "storage", "remote", "otlptranslator", "prometheusremotewrite")
+	entries, err := os.ReadDir(dpath)
+	if err != nil {
+		return nil, fmt.Errorf("list files in %q: %w", dpath, err)
+	}
+
+	var sources []string
+	for _, e := range entries {
+		if e.IsDir() || !strings.HasSuffix(e.Name(), ".go") || e.Name() == "timeseries.go" || strings.HasSuffix(e.Name(), "_test.go") {
+			continue
+		}
+
+		dst := fmt.Sprintf("%s_generated.go", e.Name()[:len(e.Name())-3])
+		if err := copyFile(filepath.Join(dpath, e.Name()), dst); err != nil {
+			return nil, err
+		}
+
+		sources = append(sources, dst)
+	}
```

Comment:
> There doesn't seem to be anything advanced here, just lot of error-checking boilerplate code

@pstibrany this is the complicated part, I'm not sure how I would rewrite it as shell script. Also, what do we gain from the rewriting effort? Are there any objective gains? I think it just becomes less understandable/maintainable.